SignalServiceKit/src/Util/OWSBackgroundTask.m

```diff
@@ -0,0 +1,157 @@
+        return self;
+    }
+
+    OWSAssert(label.length > 0);
+
+    _label = label;
+    self.completionBlock = completionBlock;
+
+    [self startBackgroundTask];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [self endBackgroundTask];
+}
+
+- (void)startBackgroundTask
+{
+    // beginBackgroundTaskWithExpirationHandler must be called on the main thread.
+    DispatchMainThreadSafe(^{
+        __weak typeof(self) weakSelf = self;
+        self.backgroundTaskId = [CurrentAppContext() beginBackgroundTaskWithExpirationHandler:^{
+            // Note the usage of OWSCAssert() to avoid capturing a reference to self.
+            OWSCAssert([NSThread isMainThread]);
+
+            OWSBackgroundTask *strongSelf = weakSelf;
+            if (!strongSelf) {
+                return;
+            }
+            @synchronized(strongSelf)
+            {
+                if (strongSelf.backgroundTaskId == UIBackgroundTaskInvalid) {
+                    return;
+                }
+                DDLogInfo(@"%@ %@ background task expired.", strongSelf.logTag, strongSelf.label);
+                strongSelf.backgroundTaskId = UIBackgroundTaskInvalid;
+
+                if (strongSelf.completionBlock) {
+                    strongSelf.completionBlock(BackgroundTaskState_Expired);
+                    strongSelf.completionBlock = nil;
+                }
+            }
+        }];
+
+        // If a background task could not be begun, call the completion block.
+        if (self.backgroundTaskId == UIBackgroundTaskInvalid) {
+
+            DDLogInfo(@"%@ %@ background task could not be started.", self.logTag, self.label);
+
+            // Make a local copy of completionBlock to ensure that it is called
+            // exactly once.
+            BackgroundTaskCompletionBlock _Nullable completionBlock;
+            @synchronized(self)
+            {
+                completionBlock = self.completionBlock;
+                self.completionBlock = nil;
+            }
+            if (completionBlock) {
+                DispatchMainThreadSafe(^{
+                    completionBlock(BackgroundTaskState_CouldNotStart);
+                });
+            }
+        }
+    });
+}
+
+- (void)endBackgroundTask
+{
+    // Make a local copy of this state, since this method is called by `dealloc`.
+    UIBackgroundTaskIdentifier backgroundTaskId;
+    BackgroundTaskCompletionBlock _Nullable completionBlock;
+    NSString *logTag = self.logTag;
+    NSString *label = self.label;
+
+    @synchronized(self)
+    {
+        backgroundTaskId = self.backgroundTaskId;
+        completionBlock = self.completionBlock;
+    }
+
+    if (backgroundTaskId == UIBackgroundTaskInvalid) {
+        OWSAssert(!completionBlock);
+        return;
+    }
+
+    // endBackgroundTask must be called on the main thread.
+    DispatchMainThreadSafe(^{
+        DDLogVerbose(@"%@ %@ background task completed.", logTag, label);
+        [CurrentAppContext() endBackgroundTask:backgroundTaskId];
```

Comment:
Should this come *after* the completion block so that we can help ensure the completion block runs before we're suspended?

(in which case you'd probably want to ensure that the completion block  is not redundantly called in the expirationHandler)

Reply:
I've juggled things around here a bit.  I've added some logic to copy/clear the `completionBlock` to ensure that it's called exactly once.